src/protocol/step/compact.rs

```diff
@@ -0,0 +1,190 @@
+
+use super::{Step, StepNarrow};
+use ipa_macros::Gate;
+use std::fmt::{Debug, Display, Formatter};
+
+#[derive(Gate, Clone, Hash, PartialEq, Eq, Default)]
+#[cfg_attr(
+    feature = "enable-serde",
+    derive(serde::Deserialize),
+    serde(from = "&str")
+)]
+pub struct Compact(pub u16);
+
+// serde::Deserialize requires From<&str> implementation
+impl From<&str> for Compact {
+    fn from(id: &str) -> Self {
+        // id begins with '/'. strip it.
+        Compact::deserialize(&id[1..])
+    }
+}
+
+impl Display for Compact {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Debug for Compact {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "step={}", self.0)
+    }
+}
+
+const ROOT_STATE: u16 = 0;
+const FALLBACK_STATE: u16 = 65535;
+const UPGRADE_SEMI_HONEST_STATE: u16 = 65534;
+const QUERY_TYPE_SEMIHONEST_STATE: u16 = 65533;
+const QUERY_TYPE_MALICIOUS_STATE: u16 = 65532;
+const PRSS_EXCHANGE_STATE: u16 = 65531;
+
+// Hard-coded state map for steps that are narrows but never executed.
+// Such steps are used in many places in the code base, either for convenience of
+// writing compact code, by design, or for tests.
+// For example, `semi_honest::Upgraded` implements `UpgradeContext` trait, but it's
+// `upgrade` method is never executed. For these steps, we need to provide the state,
+// and sub-states that branch off from it, and deliberately ignore them.
+fn static_state_map(state: u16, step: &str) -> u16 {
+    if state == ROOT_STATE && step.starts_with("run-") {
+        // ignore all `run-*` steps
+        // we use `starts_with` to avoid regex dependency. should be good enough
+        return ROOT_STATE;
+    } else if step == crate::protocol::boolean::random_bits_generator::FallbackStep.as_ref() {
+        // returning an arbitrary state here works because there are no subsequent narrows.
+        // if there were, we would need to do the same as in `UPGRADE_SEMI_HONEST_STATE` below.
+        return FALLBACK_STATE;
+    } else if step == crate::protocol::context::semi_honest::UpgradeStep.as_ref()
+        || state == UPGRADE_SEMI_HONEST_STATE
+    {
+        // if we see `upgrade_semi-honest` step, we move to UPGRADE_SEMI_HONEST_STATE to indicate
+        // that any subsequent narrows from this state will be ignored.
+        return UPGRADE_SEMI_HONEST_STATE;
+    } else if step == crate::helpers::query::QueryType::SEMIHONEST_IPA_STR {
+        return QUERY_TYPE_SEMIHONEST_STATE;
+    } else if step == crate::helpers::query::QueryType::MALICIOUS_IPA_STR {
+        return QUERY_TYPE_MALICIOUS_STATE;
+    } else if step == crate::helpers::prss_protocol::PrssExchangeStep.as_ref() {
+        return PRSS_EXCHANGE_STATE;
+    }
+
+    panic!("cannot narrow with \"{step}\" from state {state}");
+}
+
+// Reverse of `static_state_map` for `Compact::as_ref()`
+fn static_reverse_state_map(state: u16) -> &'static str {
```

Comment:
I think you rely on this function somewhere inside the macro, which makes this dependency a bit more tightly coupled that ideal. Not sure if there is a better way though

Reply:
I could move everything inside the proc macro so that the IPA core code looks a bit more cleaner, but the dependency is still there. I hope to find a better solution with a new macro but the synchronization of assigning a unique state ID for these special steps would be tricky.